File: dbflow/src/main/java/com/raizlabs/android/dbflow/structure/cache/ModelLruCache.java
package com.raizlabs.android.dbflow.structure.cache;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.raizlabs.android.dbflow.annotation.Table;

/**
 * Description: Provides an {@link com.raizlabs.android.dbflow.structure.cache.LruCache} under its hood
 * and provides synchronization mechanisms.
 */
public class ModelLruCache<TModel> extends ModelCache<TModel, LruCache<Long, TModel>> {

    /**
     * @param size The size, if less than or equal to 0 we set it to {@link Table#DEFAULT_CACHE_SIZE}.
     */
    public static <TModel> ModelLruCache<TModel> newInstance(int size) {
        if (size <= 0) {
            size = Table.DEFAULT_CACHE_SIZE;
        }
        return new ModelLruCache<>(size);
    }

    protected ModelLruCache(int size) {
        super(new LruCache<Long, TModel>(size));
    }

    @Override
    public void addModel(@Nullable Object id, @NonNull TModel model) {
        if (id instanceof Number) {
            synchronized (getCache()) {
                Number number = ((Number) id);
                getCache().put(number.longValue(), model);
            }
        } else {
            throw new IllegalArgumentException("A ModelLruCache must use an id that can cast to" +
                    "a Number to convert it into a long");
        }
    }

    @Override
    public TModel removeModel(@NonNull Object id) {
        TModel model;
        if (id instanceof Number) {
            synchronized (getCache()) {
                model = getCache().remove(((Number) id).longValue());
            }
        } else {
            throw new IllegalArgumentException("A ModelLruCache uses an id that can cast to" +
                    "a Number to convert it into a long");
        }
        return model;
    }

    @Override
    public void clear() {
        synchronized (getCache()) {
            getCache().evictAll();
        }
    }

    @Override
    public void setCacheSize(int size) {
        getCache().resize(size);
    }

    @Override
    public TModel get(@Nullable Object id) {
        if (id instanceof Number) {
            return getCache().get(((Number) id).longValue());
        } else {
            throw new IllegalArgumentException("A ModelLruCache must use an id that can cast to" +
                    "a Number to convert it into a long");
        }
    }
}

File: dbflow/src/main/kotlin/com/raizlabs/dbflow5/query/cache/ModelLruCache.kt
package com.raizlabs.dbflow5.query.cache

import android.util.LruCache
import com.raizlabs.dbflow5.annotation.DEFAULT_CACHE_SIZE

/**
 * Description: Provides an [android.util.LruCache] under its hood
 * and provides synchronization mechanisms.
 */
class ModelLruCache<TModel>(size: Int)
    : ModelCache<TModel, LruCache<Long, TModel>>(LruCache<Long, TModel>(size)) {

    override fun addModel(id: Any?, model: TModel) {
        if (id is Number) {
            synchronized(cache) {
                cache.put(id.toLong(), model)
            }
        } else {
            throw IllegalArgumentException("A ModelLruCache must use an id that can cast to" + "a Number to convert it into a long")
        }
    }

    override fun removeModel(id: Any): TModel? {
        if (id is Number) {
            synchronized(cache) {
                return cache.remove(id.toLong())
            }
        } else {
            throw IllegalArgumentException("A ModelLruCache uses an id that can cast to" + "a Number to convert it into a long")
        }
    }

    override fun clear() {
        synchronized(cache) {
            cache.evictAll()
        }
    }

    override fun setCacheSize(size: Int) {
        cache.resize(size)
    }

    override fun get(id: Any?): TModel? {
        return if (id is Number) {
            cache.get(id.toLong())
        } else {
            throw IllegalArgumentException("A ModelLruCache must use an id that can cast to" + "a Number to convert it into a long")
        }
    }

    companion object {

        /**
         * @param size The size, if less than or equal to 0 we set it to [DEFAULT_CACHE_SIZE].
         */
        fun <TModel> newInstance(size: Int): ModelLruCache<TModel> {
            var locSize = size
            if (locSize <= 0) {
                locSize = DEFAULT_CACHE_SIZE
            }
            return ModelLruCache(locSize)
        }
    }
}
